oak/server/rust/oak_glue/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+///
+/// # Safety
+///
+/// Caller must ensure that the memory range [config_buf, config_buf+config_len) is
+/// accessible and holds a protobuf-serialized ApplicationConfiguration message.
+#[no_mangle]
+pub unsafe extern "C" fn glue_start(
+    config_buf: *const u8,
+    config_len: u32,
+    factory: Option<NodeFactory>,
+    factory_data: usize,
+) -> u64 {
+    std::panic::catch_unwind(|| {
+        let config_data = std::slice::from_raw_parts(config_buf, config_len as usize);
+
+        let app_config = match ApplicationConfiguration::decode(config_data) {
+            Ok(c) => c,
+            Err(e) => {
+                error!("Failed to decode ApplicationConfiguration: {}", e);
+                return oak_abi::INVALID_HANDLE;
+            }
+        };
+        info!(
+            "start runtime with initial config {}.{}",
+            app_config.initial_node_config_name, app_config.initial_entrypoint_name
+        );
+        let (runtime, grpc_handle) = match oak_runtime::configure_and_run(app_config) {
+            Ok(p) => p,
+            Err(status) => {
+                error!("Failed to start runtime: {:?}", status);
+                return oak_abi::INVALID_HANDLE;
+            }
+        };
+
+        oak_runtime::node::external::register_factory(create_and_run_node);
+        info!("register oak_glue::create_and_run_node() as node factory");
+
+        let glue = Glue::new(runtime, factory, factory_data);
+        *GLUE.write().expect(R1) = Some(glue);
+        grpc_handle.0
+    })
+    .unwrap_or(oak_abi::INVALID_HANDLE)
+}
+
+#[no_mangle]
+pub extern "C" fn glue_stop() {
+    let runtime = GLUE.read().expect(R1).as_ref().expect(R2).runtime.clone();
+    warn!("stopping Rust runtime");
+    runtime.stop();
+    *GLUE.write().expect(R1) = None;
+}
+
+/// # Safety
+///
+/// The memory range [buf, buf+count*SPACE_BYTES_PER_HANDLE) must be
+/// valid.
+#[no_mangle]
+pub unsafe extern "C" fn glue_wait_on_channels(node_id: u64, buf: *mut u8, count: u32) -> u32 {
+    std::panic::catch_unwind(|| {
+        debug!(
+            "{{{}}}: wait_on_channels(buf={:?}, count={})",
+            node_id, buf, count
+        );
+
+        // Accumulate the handles we're interested in.
+        let size = oak_abi::SPACE_BYTES_PER_HANDLE * count as usize;
+        let mut handle_data = Vec::<u8>::with_capacity(size);
+        std::ptr::copy_nonoverlapping(buf, handle_data.as_mut_ptr(), size);
+        handle_data.set_len(size);
+        let mut handles = Vec::with_capacity(count as usize);
+        let mut mem_reader = Cursor::new(handle_data);
+        for _ in 0..count {
+            let handle = mem_reader.read_u64::<byteorder::LittleEndian>().unwrap();
+            let _b = mem_reader.read_u8().unwrap();
+            debug!("{{{}}}: wait_on_channels handle {:?}", node_id, handle);
+            handles.push(Some(oak_runtime::Handle(handle)));
+        }
+
+        let proxy = GLUE.read().expect(R1).as_ref().expect(R2).proxy(node_id);
```

Comment:
It seems that `GLUE.read().expect(R1).as_ref().expect(R2).proxy(node_id)` could be encapsulated in a top-level fn?